### PR TITLE
Fixing ml

### PR DESCRIPTION
**Pull request by JohnYao93**
Fixed the trace bugs
Fixed the problem of not having an angle at the end of a trace
Added capabilities to train an ml model on multiple types of maps
LSTM is training right now in the college computing cluster

TODO: merging VIN and MPNet into PB

### CODE REVIEW

File: src/algorithms/lstm/ML_model.py [hunk 1]
```diff
@@ -169,7 +169,8 @@ def __init__(self, seq: List[torch.Tensor]) -> None:
         self.data, self.lengths = pad_packed_sequence(packed_sequence, batch_first=True)
 
     def __getitem__(self, index: int) -> Tuple[Any, int]:
-        return self.data[index], self.lengths[index]
+        d = self.data[index], self.lengths[index]
```
Comment (judicaelclair) [inline src/algorithms/lstm/ML_model.py:172]: Uhm why do you assign this to a temporary?

File: src/main.py [hunk 2]
```diff
@@ -121,6 +124,7 @@ def main() -> bool:
                                      formatter_class=argparse.RawTextHelpFormatter)
     parser.add_argument("-v", "--visualiser", action='store_true', help="run simulator with graphics")
     parser.add_argument("-g", "--generator", action='store_true', help="run generator")
+    parser.add_argument("--trainer", action='store_true', help="runs model trainer")
```
Comment (judicaelclair) [inline src/main.py:127]: You could add a `-t` option

File: src/simulator/services/algorithm_runner.py [hunk 3]
```diff
@@ -143,8 +143,8 @@ def find_path(self, agent_pos: Point = None, goal_pos: Point = None) -> None:
         self.reset_algorithm()
         if self._services.settings.simulator_key_frame_speed > 0:
             self.instance.set_root_key_frame(self.__pass_root_key_frame)
-        if agent_pos:
+        if agent_pos is not None:
```
Comment (judicaelclair) [inline src/simulator/services/algorithm_runner.py:146]: Why change this? I prefer having explicit `None` check.

File: src/simulator/services/algorithm_runner.py [hunk 4]
```diff
@@ -143,8 +143,8 @@ def find_path(self, agent_pos: Point = None, goal_pos: Point = None) -> None:
...
             self.map.move_agent(agent_pos, True)
-        if goal_pos:
+        if goal_pos is not None:
```
Comment (judicaelclair) [inline src/simulator/services/algorithm_runner.py:148]: Dido

File: src/simulator/services/resources/directory.py [hunk 5]
```diff
@@ -67,7 +67,8 @@ def load(self, name: str, load_function: Callable[['Directory', str], Any] = Non
             self._services.debug.write("Loaded [{}]".format(self._full_path() + name), DebugLevel.LOW)
         else:
             self._services.debug.write("File not found [{}]".format(self._full_path() + name), DebugLevel.LOW)
-            raise RuntimeError("File not found, consider running from src/")
+            print(self._full_path() + name)
```
Comment (judicaelclair) [inline src/simulator/services/resources/directory.py:70]: take this print statement out and ideally restore the exception. If you're debugging set the debug level high to get output

**Review by judicaelclair — CHANGES_REQUESTED** [review]

File: src/simulator/services/resources/directory.py (hunk 5 above)
Comment (judicaelclair) [reply src/simulator/services/resources/directory.py:70]: Consider adding a debug flag to `main.py` with optionally specified level. If level not specified goes to highest possible.

**Review by judicaelclair — APPROVED** [review]

File: src/algorithms/basic_testing.py [hunk 6]
```diff
@@ -196,7 +196,8 @@ def get_occupancy_percentage_size(size: Size, items_nr: int) -> float:
 
     @staticmethod
     def get_euclidean_distance_traveled(trace: List[Trace], agent: Agent) -> float:
-        trace.append(Trace(agent.position))
+        if (not trace) or trace[-1].position != agent.position:
```
Comment (Nabushika) [inline src/algorithms/basic_testing.py:199]: can we change `(not trace)` to `len(trace) == 0`?

File: src/algorithms/lstm/map_processing.py [hunk 7]
```diff
@@ -182,8 +182,8 @@ def agent_goal_angle_feature(mp: Map) -> torch.Tensor:
     @staticmethod
     def valid_moves_feature(mp: Map) -> torch.Tensor:
         neighbours: Set[Point] = set(mp.get_neighbours(mp.agent.position))
-        res: torch.Tensor = torch.zeros(8)
-        for idx, mv in enumerate(mp.EIGHT_POINTS_MOVE_VECTOR):
+        res: torch.Tensor = torch.zeros(8) if mp.size.n_dim == 2 else torch.zeros(26)
```
Comment (Nabushika) [inline src/algorithms/lstm/map_processing.py:185]: We could change this to `len(ALL_POINTS_MOVE_VECTOR)` to be more general

File: src/main.py [hunk 8]
```diff
@@ -107,9 +107,25 @@ def configure_and_run(args) -> bool:
...
+    elif args.debug == 'BASIC':
+        config.simulator_write_debug_level = DebugLevel.BASIC
+    elif args.debug == 'LOW':
+        config.simulator_write_debug_level = DebugLevel.LOW
+    elif args.debug == 'MEDIUM':
+        config.simulator_write_debug_level = DebugLevel.MEDIUM
+    elif args.debug == 'HIGH':
+        config.simulator_write_debug_level = DebugLevel.HIGH
+    else:
+        raise NotImplementedError
```
Comment (Nabushika) [inline src/main.py:128]: Could we do this with a dictionary? Might be neater, e.g.

```
debug_levels = {"NONE": DebugLevel.None, ...}
config.simulator_write_debug_level = debug_levels[args.debug]
```

Comment (judicaelclair) [reply src/main.py:128]: Better even would be to iterate over the debug levels using the `DebugLevel.[..].__name__` using list comprehension

**Pull request closed without merging**